src/createCompositor.js

```diff
@@ -289,21 +293,57 @@ function createSizeParser(name, exprFunc, scale) {
     };
 }
 
-function tryConvertExpr(string) {
-    // PENDING
-    var exprRes = /^expr\((.*)\)$/.exec(string);
-    if (exprRes) {
-        try {
-            var func = new Function('width', 'height', 'dpr', 'return ' + exprRes[1]);
-            // Try run t
-            func(1, 1);
+function astToMathExpression(node, context) {
+    switch (node.type) {
+        case 'BinaryExpression':
+            let left = astToMathExpression(node.left, context);
+            let right = astToMathExpression(node.right, context);
+            return `(${left} ${node.operator} ${right})`;
+        case 'Literal':
+            return node.value;
+        case 'Identifier':
+            if (['width', 'height', 'dpr'].includes(node.name)) {
+                return context[node.name];
+            }
+            throw new Error(`Unknown identifier: ${node.name}`);
+        default:
+            throw new Error(`Unsupported node type: ${node.type}`);
+    }
+}
 
-            return func;
-        }
-        catch (e) {
-            throw new Error('Invalid expression.');
+function getSafeExpressionEvaluator(expression) {
+    if (!expression) return;
+    var exprRes = /^expr\((.*)\)$/.exec(expression);
+    if (!exprRes) return;
+
+    const ast = esprima.parseScript(exprRes[1]);
+
+    const func = function(width, height, dpr) {
+        const context = { width, height, dpr };
+        const mexp = new Mexp()
```

Comment:
Can the `const mexp` value move out of the `func` scope so in instantiated for each call?